### PR TITLE
bpo-30687: Fixes build scripts to find msbuild.exe and stop relying on vcvarsall.bat

### PR DESCRIPTION
**Pull request by zooba**
Fixes build scripts to find msbuild.exe and stop relying on vcvarsall.bat
Also fixes bdist_wininst.vcxproj to use correct version in generated name.

### CODE REVIEW

**Comment by zooba** [comment]
Decided this isn't really trivial, so I'll create an issue. Not a lot to discuss right now I don't think, but there could be when this all goes wrong :)

**Comment by zooba** [comment]
@tjguk This should fix up the trouble you were having earlier 👍

File: PCbuild/fix_lineendings.py
Comment (zware) [inline PCbuild/fix_lineendings.py]: This isn't right, they should be CRLF :).  Also, a `del .git\index && git reset` from a clean (no changes, no untracked or ignored files) checkout should give you proper line endings according to the rules in `.gitattributes`.

File: PCbuild/find_msbuild.bat
Comment (zware) [inline PCbuild/find_msbuild.bat:11]: Does this work, or should these be double quotes?  I've been bitten by trying to use single quotes in batch files a couple of times, but I'm not convinced that I actually understand the quoting rules.

File: PCbuild/find_msbuild.bat
Comment (zware) [inline PCbuild/find_msbuild.bat]: VS7?

File: PCbuild/build.bat
Comment (zware) [inline PCbuild/build.bat:118]: Can `vs_platf` go away entirely now?

**Review by zware — COMMENTED** [review]
I like the idea.  Haven't tested it yet, and have a few questions, below.

File: PCbuild/fix_lineendings.py
Comment (zooba) [reply PCbuild/fix_lineendings.py]: Well, they weren't CRLF, because Mercurial (I'm using hggit on this machine) fixed them all and they all got changed.

I can fix them back, but then the review is going to be unreadable, so check it all out before I do it :)

File: PCbuild/find_msbuild.bat
Comment (zooba) [reply PCbuild/find_msbuild.bat:11]: The quoting rules are horrible here, because I want to handle where users have done:

```
set MSBUILD=C:\path to\MSBuild.exe
set MSBUILD="C:\path to\MSBuild.exe"
```

Which breaks in a whole lot of ridiculous ways. Using single quotes here at least handles the two sane cases without causing syntax errors.

File: PCbuild/find_msbuild.bat
Comment (zooba) [reply PCbuild/find_msbuild.bat]: Yep. One of those names that stuck for back-compat reasons...

File: PCbuild/build.bat
Comment (zooba) [reply PCbuild/build.bat:118]: Yep, sure can!

Comment (zooba) [reply PCbuild/build.bat:118]: But then I think the cross-compilation for PGO warning has to go away too... not sure of the full impact there yet, let me try it.

Comment (zware) [reply PCbuild/build.bat:118]: It shouldn't; the `vs_platf` stuff just gets removed from around it.

File: PCbuild/fix_lineendings.py
Comment (zware) [reply PCbuild/fix_lineendings.py]: Hmm.  I just checked it out and got the right line endings even on your new `find_msbuild.bat` file.  I think this file can just be removed and we hope `.gitattributes` makes Git do the right thing :)

File: PCbuild/build.bat
Comment (zooba) [reply PCbuild/build.bat:118]: Looks like there are some missing environment pieces, but I can add those into `python.bat` on build. Looks like that's how we launch for PGO build anyway.

File: PCbuild/fix_lineendings.py
Comment (zooba) [reply PCbuild/fix_lineendings.py]: I needed the script to fix up the problems when hggit did the wrong thing. Maybe I just need to switch to git completely (though then I'm *never* going to backport anything, because it takes 10x as long)

File: PCbuild/build.bat
Comment (zooba) [reply PCbuild/build.bat:118]: Ah, I see what you mean about keeping the warning in there. I'll bring that back

**Comment by zooba** [comment]
@ned-deily Can we consider this for 3.6.2? It has no impact on the resulting build, but will make the sdist (or forks) easier to use on Windows.

**Comment by ned-deily** [comment]
I trust your judgement on the impact and importance.  As you know, 3.6.2rc1 is already being "manufactured".  Since this is more than a trivial change and presumably could have end-user impact, I don't think it would be appropriate to try to drop it into a 3.6.2final without exposure.  There are a couple of options we can discuss off-line but, in any case, the change needs to first be merged into the normal 3.6 branch.

**Comment by zooba** [comment]
I'm just waiting on @zware to approve (and merge, if there are no changes needed). The backport to the 3.6 branch should be trivial, and probably 3.5 as well.

**Comment by tjguk** [comment]
I'm looking at this but it's likely I won't have a chance to test 
tonight (9.30pm my time now). Obviously if @zware can test it out, then 
great!

**Comment by tjguk** [comment]
Ok; I have tested and confirm the build now works for VS2017. I haven't 
reviewed the code changes as such, and I presume that @zware is a more 
appropriate reviewer in any case.

File: PCbuild/build.bat
Comment (zware) [inline PCbuild/build.bat]: Error message needs to be reworded, since we're no longer checking %ProgramFiles(x86)% (and %on_64_bit% should be removed from above `:CheckOpts`)

File: PCbuild/fix_lineendings.py
Comment (zware) [reply PCbuild/fix_lineendings.py]: Would `Tools/scripts/crlf.py` be enough for you?

File: PCbuild/find_msbuild.bat
Comment (zware) [inline PCbuild/find_msbuild.bat]: I think it's here that I get `ERROR: The system was unable to find the specified registry key or value.` on a machine without VS 2017.  Can that be suppressed without too much hardship?  Probably the same below, but that key was actually found on my machine.

File: PCbuild/find_msbuild.bat
Comment (zware) [inline PCbuild/find_msbuild.bat]: Since the idea is to mutate the enclosing environment, `_SOURCE` should probably be something even less likely to conflict, like `_Py_MSBuild_Source`.

**Review by zware — CHANGES_REQUESTED** [review]
One change needed, one suggestion, one request, and one question.

**Comment by zooba** [comment]
Made the requested changes, now just working on dealing with line endings...

**Review by zware — APPROVED** [review]

**Comment by zooba** [comment]
I think the line endings problem is that .hgeol converts on update/checkout but not commit, while git converts everything to LF on commits and back on checkout. So the actual files in the repo are LF and become CRLF on checkout, but hggit puts them back into git as CRLF instead of LF and we get stupid diffs.

Once we get a backport bot I'll have less need to use something sensible (i.e. hg) for backporting and can use git for more development. Until then, guess I'll just keep doing these extra commits - luckily they'll squash out.

**Comment by zware** [comment]
What kind of behavior do you get if you just remove `.hgeol`?

**Comment by zooba** [comment]
> remove `.hgeol`

Then all my .bat files have LF endings. I'm on Win10 everywhere, so that's less of an issue for me - I think it's only Win7 and earlier that don't like them.

**Comment by tjguk** [comment]
On 19/06/2017 18:25, Steve Dower wrote:
> Once we get a backport bot I'll have less need to use something sensible
> (i.e. hg) for backporting and can use git for more development. Until
> then, guess I'll just keep doing these extra commits - luckily they'll
> squash out.

Makes it harder to review within GH though -- have to manually scan the 
every-line-has-changed looking for actual changes!

**Comment by zooba** [comment]
> Makes it harder to review within GH though 

Yep, that's why I have about 8 "Fix line endings" commits, to get around that problem :)

**Comment by tjguk** [comment]
:)

**Pull request merged**
